src/topology/continuous_function/stone_weierstrass.lean

```diff
@@ -0,0 +1,353 @@
+  intro z,
+
+  -- We rewrite into this particular form,
+  -- so that simp lemmas about inequalities involving `finset.inf'` can fire.
+  rw [(show ∀ a b ε : ℝ, dist a b < ε ↔ a < b + ε ∧ b - ε < a,
+    by { intros, simp only [← metric.mem_ball, real.ball_eq_Ioo, set.mem_Ioo, and_comm], })],
+
+  fsplit,
+  { dsimp [k],
+    simp only [finset.inf'_lt_iff, continuous_map.inf'_apply],
+    obtain ⟨x, xm, zm⟩ := set.exists_set_mem_of_union_eq_top _ _ xs_w z,
+    exact ⟨x, xm, h_lt _ _ zm⟩, },
+  { dsimp [k],
+    simp only [finset.lt_inf'_iff, continuous_map.inf'_apply],
+    intros x xm,
+    apply lt_h, },
+end
+
+/--
+The Stone-Weierstrass approximation theorem,
+that a subalgebra `A` of `C(X, ℝ)`, where `X` is a compact topological space,
+is dense if it separates points.
+-/
+theorem subalgebra_topological_closure_eq_top_of_separates_points
+  (A : subalgebra ℝ C(X, ℝ)) (w : A.separates_points) :
+  A.topological_closure = ⊤ :=
+begin
+  -- The closure of `A` is closed under taking `sup` and `inf`,
+  -- and separates points strongly (since `A` does),
+  -- so we can apply `sublattice_closure_eq_top`.
+  apply set_like.ext',
+  let L := A.topological_closure,
+  have n : set.nonempty (L : set C(X, ℝ)) :=
+    ⟨(1 : C(X, ℝ)), A.subalgebra_topological_closure A.one_mem⟩,
+  convert sublattice_closure_eq_top
+    (L : set C(X, ℝ)) n
+    (λ f g fm gm, inf_mem_closed_subalgebra L A.is_closed_topological_closure ⟨f, fm⟩ ⟨g, gm⟩)
+    (λ f g fm gm, sup_mem_closed_subalgebra L A.is_closed_topological_closure ⟨f, fm⟩ ⟨g, gm⟩)
+    (subalgebra.separates_points.strongly
+      (subalgebra.separates_points_monotone (A.subalgebra_topological_closure) w)),
+  { simp, },
+  { ext, simp, },
+end
+
+/--
+An alternative statement of the Stone-Weierstrass theorem.
+
+If `A` is a subalgebra of `C(X, ℝ)` which separates points (and `X` is compact),
+every real-valued continuous function on `X` is a uniform limit of elements of `A`.
+-/
+theorem continuous_map_mem_subalgebra_closure_of_separates_points
+  (A : subalgebra ℝ C(X, ℝ)) (w : A.separates_points)
+  (f : C(X, ℝ)) :
+  f ∈ A.topological_closure :=
+begin
+  rw subalgebra_topological_closure_eq_top_of_separates_points A w,
+  simp,
+end
+
+/--
+An alternative statement of the Stone-Weierstrass theorem,
+for those who like their epsilons.
+
+If `A` is a subalgebra of `C(X, ℝ)` which separates points (and `X` is compact),
+every real-valued continuous function on `X` is within any `ε > 0` of some element of `A`.
+-/
+theorem exists_mem_subalgebra_near_continuous_map_of_separates_points
+  (A : subalgebra ℝ C(X, ℝ)) (w : A.separates_points)
+  (f : C(X, ℝ)) (ε : ℝ) (pos : 0 < ε) :
+  ∃ (g : A), ∥(g : C(X, ℝ)) - f∥ < ε :=
+begin
+  have w := mem_closure_iff_frequently.mp
+    (continuous_map_mem_subalgebra_closure_of_separates_points A w f),
+  rw metric.nhds_basis_ball.frequently_iff at w,
+  obtain ⟨g, H, m⟩ := w ε pos,
+  rw [metric.mem_ball, dist_eq_norm] at H,
+  exact ⟨⟨g, m⟩, H⟩,
+end
+
```

Comment:
Can you also give the statement for those who like their epsilons and their unbundled continuous functions?

Reply:
Thanks very much for these suggestions, Sebastien! I did think about adding a statement for unbundled continuous functions here after your suggestion on the `weierstrass` branch, but I'm actually not sure what to say. It seems I would have to avoid even using `subalgebra`, and that seems like it is getting pretty cumbersome.

Reply:
I would still use `subalgebra`, as in `exists_mem_subalgebra_near_continuous_map_of_separates_points`, but take a function `f`, assume `continuous f`, and deduce the existence of `g` in the subalgebra which is pointwise close to `f`.

Reply:
Done!